### PR TITLE
Tracing: fix explicit context propagation when running in javaagent

### LINKED ISSUE #38221: OTel: Document disabling reactor plugin and verify that manual context propagation works with agent
We need to document (and may be even recommend) disabling reactor context propagation due to perf overhead and polluted stack traces.

We also need to check if [explicit context propagation](https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/core/azure-core-tracing-opentelemetry) works with agent - I believe it does not due to shading in the agent.

### PR DESCRIPTION
**Pull request by lmolkova**
Fixes #38221

See https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/11042 for the context.

TL;DR:
when azure-core-tracing-otel is running inside Java OTel or ApplicationInsights agents, the otel context is relocated.
However users passing context explicitly pass the original non-relocated context. 
As a result, we can't correlate with the explicit context passed by users.

We need to handle both cases.
This change detects if we're running in the agent and converts original context to the relocated one using otel-provided public APIs using reflection.

#38221 also lists documentation, which we already have in the 
https://github.com/Azure/azure-sdk-for-java/blob/2f56b2187def966f890da45136913065780e5cac/sdk/core/azure-core-tracing-opentelemetry/README.md?plain=1#L104-L106

and 

https://github.com/Azure/azure-sdk-for-java/blob/2f56b2187def966f890da45136913065780e5cac/sdk/core/azure-core-tracing-opentelemetry/README.md?plain=1#L149-L152

### CODE REVIEW

**Comment by lmolkova** [comment]
I will add a test to validate it inside the [shaded plugin](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/azure-core/azure-core-1.36/javaagent/src/testAzure/java/io/opentelemetry/javaagent/instrumentation/azurecore/v1_36/AzureSdkTest.java)

File: sdk/core/azure-core-tracing-opentelemetry/src/main/java/com/azure/core/tracing/opentelemetry/OpenTelemetryTracer.java [hunk 1]
```diff
@@ -51,6 +52,38 @@ public class OpenTelemetryTracer implements com.azure.core.util.tracing.Tracer {
...
+        try {
+            Class<?> agentCtxStorageClass = Class
+                .forName("io.opentelemetry.javaagent.instrumentation.opentelemetryapi.context.AgentContextStorage");
+
+            // Shading plugins rewrite string literals that look like package name, so we have to trick it
+            // to resolve the actual io.opentelemetry.context.Context.
+            // (If we're running in agent, this code was rewritten to use io.opentelemetry.javaagent.shaded.io.opentelemetry.context.Context)
+            applicationContextClass = Class.forName("#io.opentelemetry.context.Context".substring(1));
+            convertAppToAgentContextMethod = agentCtxStorageClass.getMethod("getAgentContext", applicationContextClass);
+        } catch (Throwable t) {
```
Comment (alzimmermsft) [inline sdk/core/azure-core-tracing-opentelemetry/src/main/java/com/azure/core/tracing/opentelemetry/OpenTelemetryTracer.java:82]: Should we check for `t` being either a reflection exception or error and logging otherwise throwing

**Review by alzimmermsft — APPROVED** [review]

**Review by trask — APPROVED** [review]
👍
